### ISSUE TITLE
"Publish" Button gets stuck while trying to comment on PR done from and to forked repo

### ISSUE THREAD

**Issue by artiomle** [Reviewable/Reviewable#735]
While trying to comment on PR form forked repo branch to organization branch all works well

While trying to comment on PR from forked repo branch to same forked repo branch - the Submit button gets stuck (the changes only saved as drafts)

**Comment by artiomle** [comment]
We are using reviewable version 2183.3776
GitHub enterprise version is 2.17.9

**Comment by pkaminski** [comment]
That sounds odd.  When you say "Submit button" do you mean Publish or Merge?  "Stuck" means the button remains disabled but no error shows up?  Is there anything in the console?

**Comment by artiomle** [comment]
I mean Publish button. It's green and clickable, but when clicking on it, the pending animation starting (round spinning) and nothing happens. The container console log has no error. In chrome f12 dev console, no beckend request is sent (no sure if you are using web socket though).On 21 May 2020 23:29, Piotr Kaminski <notifications@github.com> wrote:
That sounds odd.  When you say "Submit button" do you mean Publish or Merge?  "Stuck" means the button remains disabled but no error shows up?  Is there anything in the console?

**Comment by pkaminski** [comment]
Thanks for the details.  One more thing:  you're running a version that's over 8 months out of date.  Could you please update to the latest (2330.4029) and see if the issue still repros?  Thanks.

**Comment by artiomle** [comment]
I copied the wrong version in the original issue comment. I'm actually running the latest version (2330.4029). Sorry for the confusion.

**Comment by pkaminski** [comment]
@badz0 could you try to repro this when you have a moment please?  Thanks.

**Comment by artiomle** [comment]
I think it worth mentioning that the issue is only exist for non owner users. For example, I'm the owner of the license of reviewable and the issue is not recreated for me. Other users trying to publish comment on PR on forked repo (different branches) the button just do nothing...

**Comment by badz0** [comment]
Unfortunately, I couldn't reproduce it.

**Comment by pkaminski** [comment]
I wasn't able to reproduce the issue either.  @artiomle could you please provide more details that might be affecting things?  Of particular interest:
- Does this happen for private, public, or internal repos?
- What access level do the affected users have to the repo?  (pull, push, admin)
- Are there any custom review completion rules, either on the original or the forks?
- Do the comments you're trying to publish have any @-mentions or inline assignee/reviewer/label directives?
- What disposition (comment/approve/request changes) are you publishing with?  Is "sync requested reviewers" checked?  Is the user publishing the author of the PR, or a reviewer?
- Are you able to send individual comments using the "send immediately" button in the lower-right corner of a draft?
- Is the forked repo connected to Reviewable (on the Repositories page) or is the review ad-hoc?
- What happens if you wait 30-60 seconds after hitting Publish?  (I want to make sure any timeouts have a chance to raise an error.)
- Does anything unusual show up in the server logs when publishing?  (Again, wait up to 30-60 seconds, in case an exception log gets sent from the client to the server.)

**Comment by artiomle** [comment]
1) Private repo 
2) Write access level 
3) custom rules - Not sure - how do I check if we do (where is it configured?)
4) no - any comment gets stuck (even single character)
5) comment , "sync requested reviewers" not checked , author.
6) "send immediately" also not working (took me a while to find the button - no hover text..) 
7) yes - forked repo also connected to Reviewable
8+9) northing changes in 120 seconds 

New information (important) !!

During the info gather I had to click the arrow near the Publish button in order to see if "sync requested reviewers" is checked - at that point it gets stuck (before clicking publish) the problem is that the "preview" is gets stuck (on front-end) and while its loading (for ever) the publish button is grayed out and not clickable - this is a front-end issue that is why the container doesn't have any logs. The publish request never not even reaches the server but get's stuck on client. 

Hope that will narrow down your search for the bug

**Comment by pkaminski** [comment]
Thanks for the extra information -- that does narrow things down a bit, but not enough yet to form a good hypothesis.  Since sending a single comment fails as well (I assume the button just gets disabled and never comes back?) could you please try one more thing:  send a trivial comment like that, then check the contents of `Raven._breadcrumbs` from the JS console.  The sending process is somewhat instrumented, and the last few entries in the breadcrumbs array should give us a rough idea of where it's getting stuck.  Thanks!

**Comment by KassHunters** [comment]
Hey, @pkaminski and thanks for your help!

Breadcrums show 4 initial loading events.:
 - Viewed Review Page
 - Download the Vue Devtools extension for a better development experience
 - Connected
 - Auth

At this point everything seems stable and the webhooks reach the repo.

Here are the added breadcrums for each action we're trying:
- Publish a specific comment:
```
{
	"timestamp": 1591181170.056,
	"category": "ui.click",
	"message": "i.external.share.icon[alt=\"send message\"]"
}
```
- Publish all comments:
```
{
	category: "ui.click"
	message: "span.ui.dimmable.green.publish.buttons > div.ui.button > i.external.share.icon"
	timestamp: 1591181496.357
}
```
- Hover over publish button, and expand it (no publish action):
```
{
	category: "ui.click"
	message: "span.ui.button.green.icon > i.caret.down.icon[alt="preview published message"]"
	timestamp: 1591181615.605
}
```
None of the above are ever received as webhook in our repository.

The problem seems to be client side (even preview fails), persistent & reproducible (we've re-created it across users and repositories)

**Comment by pkaminski** [comment]
Thanks for taking a look at the breadcrumbs!  But what I'm really
interested in is whether any breadcrumbs show up _after_ you click a button
-- things like {category: "process", message: "Publishing comments"} and
similar.  Also, if you load the page with `?debug=model` appended, there
will be extra entries in the breadcrumbs that describe all database writes
attempted by the client.  All those should help me put bounds on where the
process is getting stuck.  (Note that some of the breadcrumbs may hold data
that you don't want to make public -- if so, please redact or send directly
to support@reviewable.io.)  Thanks!

**Comment by N-Kass** [comment]
Hey. Thank you for your response!
The previous comment included breadcrumbs after clicking both the top "Publish" button & the comment-specific "Send Message" button.

In debug, before pressing a button, after only entering text into a comment section, breadcrumbs have 14 events. Pressing the main "Publish" button adds just one additional event: 
```
{
	"timestamp": 1591969731.282,
	"category": "ui.click",
	"message": "span.ui.dimmable.green.publish.buttons > div.ui.button"
}
```
Even after 2 minutes, this is the only event that happens after clicking the button (Publish button still shows loading).


Still in debug, same as before, pressing the comment-specific "Send Message" button adds just one additional event: 
```
{
	"timestamp": 1591970012.395,
	"category": "ui.click",
	"message": "i.external.share.icon[alt=\"send message\"]"
}
```
This too, hangs loading after 2 minutes with no further events in breadcrumbs.


Here is the full breadcrumbs object. After redacting some info.
```
[
	{
		"timestamp": 1591969971.826,
		"category": "action",
		"message": "Viewed Review Page",
		"data": {
			"title": "Pull Request Review",
			"windowWidth": 982,
			"windowHeight": 722,
			"debug": "model",
			"pullRequest": "6",
			"build": 4029
		}
	},
	{
		"timestamp": 1591969971.9,
		"message": "Download the Vue Devtools extension for a better development experience:\nhttps://github.com/vuejs/vue-devtools",
		"level": "info",
		"category": "console"
	},
	{
		"timestamp": 1591969971.919,
		"category": "connection",
		"message": "Disconnected"
	},
	{
		"timestamp": 1591969972.837,
		"message": "set /queues/requests/-M9cnV8Z6KlwB3F68y1O [object Object]",
		"level": "log",
		"category": "console"
	},
	{
		"timestamp": 1591969972.862,
		"category": "auth",
		"data": {
			AUTH_DATA 
			"github": {
				"scopes": [
					"admin:repo_hook",
					"read:org",
					"repo",
					"user:email",
					"write:repo_hook",
					"read:repo_hook",
					"public_repo",
					"repo_deployment",
					"repo:status",
					"notifications"
				],
				"baseScopes": [
					"admin:repo_hook",
					"read:org",
					"repo",
					"user:email"
				],
				"baseScopesString": "admin:repo_hook,read:org,repo,user:email"
			}
		}
	},
	{
		"timestamp": 1591969972.865,
		"category": "warden",
		"message": "Pinged permission ticket /queues/permissions/USER|REPO|github:64"
	},
	{
		"timestamp": 1591969972.866,
		"message": "update /queues/permissions/USER|REPO|github:64 [object Object]",
		"level": "log",
		"category": "console"
	},
	{
		"timestamp": 1591969973.181,
		"category": "connection",
		"message": "Connected"
	},
	{
		"timestamp": 1591969973.715,
		"message": "update /queues/githubPullRequestSync/USER|REPO|6|github:64 [object Object]",
		"level": "log",
		"category": "console"
	},
	{
		"timestamp": 1591969973.765,
		"message": "set /queues/responses/-M9cnV8Z6KlwB3F68y1O null",
		"level": "log",
		"category": "console"
	},
	{
		"timestamp": 1591969974.917,
		"message": "set /reviews/-M9Ugkz6pLo_B9JH6TvL/participants/github:64/draftsPending true",
		"level": "log",
		"category": "console"
	},
	{
		"timestamp": 1591969974.971,
		"message": "update /queues/reconciliation/USER|REPO|6 [object Object]",
		"level": "log",
		"category": "console"
	},
	{
		"timestamp": 1591969975.757,
		"message": "set /queues/responses/-M9cnUgJ6erKrRAlHwmF null",
		"level": "log",
		"category": "console"
	},
	{
		"timestamp": 1591969991.483,
		"category": "ui.click",
		"message": "textarea#draft-top.draft.display.textarea.mp-sensitive"
	},
	{
		"timestamp": 1591970012.395,
		"category": "ui.click",
		"message": "i.external.share.icon[alt=\"send message\"]"
	}
]
```

**Comment by pkaminski** [comment]
Thanks for gathering this data -- it's most unexpected.  To begin with, I don't see the draft being saved to the database, but since you don't complain about drafts disappearing (even after a failed send) I assume the write must be happening at some point.  If you create a new draft (with `?debug=model` set), do you see messages similar to `update /users/github:64/state/-M9Ugkz6pLo_B9JH6TvL/drafts/<some id>`?  I'm guessing the breadcrumb trace you supplied had a draft already filled out and saved on a previous page load.

If so, and given that there's no further breadcrumbs appearing, that leaves only 3 suspect operations:  checking for unposted drafts, rendering the drafts' markdown, and analyzing directives.  If either of the first two were to block you'd see HTTP requests to GHE in the Network debugger -- can you confirm whether any requests show up there after you click send?

Again, since the original report said there's no requests to GHE, let's focus on the last operation.  Analyzing directives (even if none are given) requires some data to be successfully loaded from GitHub that is specific to the repo, so there's a chance that a forked repo is somehow misbehaving here.  Substituting your own organization and repo names (any case) for `owner` and `repo` respectively (and running with `?debug=model`), could you please see what the following expression return after trying to send:
1.  `truss.store.services.completions.labels.get(owner, repo)`
2.  `truss.store.services.completions.milestones.get(owner, repo)`
3.  `truss.store.services.completions.assignees.get(owner, repo)`
4.  `truss.store.services.completions.people.getCollaborators(owner, repo)`

If any of these return `undefined` or `null`, could you also double-check that they return a non-nil value in a "normal" repo, and whether there are any failed requests to GitHub in the Network debugger at any point when loading the page?

Sorry for making you do all this debugging work, but I can't think of any other way to get to the bottom of this!

**Comment by N-Kass** [comment]
You're right. Last comment was done with a draft. Creating a new message creates 3 events in breadcrumbs. 'update' to the draft state, 'set' to review last update time and participant & 'set' to the user draft.

There seems to be a request not seen before, to our GHE markdown api. This isn't always sent, 
and returns 200 with what seems to be valid data.

Per your request. I checked the values in truss, and out of 4 checks, 3 returned a value. Only the last, 'getCollaborators') returned undefined. The weird thing is I send a comment, in debug, in a different repo and all 4 requests return undefined. 
FYI, in the forked repo, just before pressing the publish, a GET request to 'collaborators' returned a list of valid collaborators.
There are no failed requests on load, pre nor post publish besides a 404 to an emoji png

This is the second forked repo we've encountered this issue with and it seems replicable. Maybe we should try and help you reproduce this issue.

**Comment by pkaminski** [comment]
Thanks again for working through this.  `getCollaborators` returning `undefined` is almost certainly what's blocking the send (and preview) from completing.  I suspect that all 4 being `undefined` in a (working) repo is a false positive of some sort, since nothing would work if that was really the case.  Let's ignore that as a red herring.

What doesn't make sense is that if the GET for `collaborators` is succeeding then there's nothing else standing in the way of computing the value that `getCollaborators` returns.  Perhaps you could dig one level deeper:  after trying to send (or preview, and as always in `?debug=model` mode), check the contents of `truss.store.services.completions.people._collaborators`.  Is there anything in the collection, and if so at what key?  (Note that this object will have a somewhat unusual configuration with property getters and such, so you might need to poke around a bit to get at its real contents.)

I also agree that we should take another swing at making a repro.  Here's what I tried:
1.  Start with a public repo in an organization.
2.  Switch to a user who only has `pull` permissions on the repo.
3.  Make a personal fork of the repo.
4.  Edit a file in the fork and create a branch.
5.  Create a PR in the fork to merge the newly created branch into `master` (still within the fork).
6.  Create a review and send a comment via Reviewable.

Are there any differences between this procedure and what you guys are doing?

One caveat is that I'm trying this on github.com rather than a GHE instance but it seems unlikely that there'll be a material difference in behavior.  Just in case, though, what version of GHE are you running?

**Comment by N-Kass** [comment]
Thanks again. Sounds like you did exactly what we did. But we (I'm with @artiomle)
are using GHE version 2.17.9 so lets try it on that

The collaborators collection doesn't change after pressing a button.
The collection contains 1 key - "<OUR_ORG_NAME>|<REPO_NAME>".
The collection didn't contain a key for the repo under my user

**Comment by pkaminski** [comment]
OK, we'll try it on GHE.  But you may have uncovered a big hint in the contents of the collaborators collection:  what is the relationship of the repo key it contains to the forked repo?  Is the GET for `collaborators` that you observe for this repo, or for the forked one?  (This is assuming that you're loading the review page "fresh" so there's not cached data from other reviews stuck in the collection.)  Perhaps Reviewable is fetching collaborators for the wrong repo (due to the fork?) and then waiting forever for the actual repo's collaborators to appear.

**Comment by N-Kass** [comment]
The GET for collaborators is for the main repo, not the forked one. And it contains all of our collaborators.

**Comment by pkaminski** [comment]
Is the main repo also the one that shows up in the `_collaborators` collection?

**Comment by N-Kass** [comment]
Yes.

**Comment by pkaminski** [comment]
So, it appears that for you trying to GET the forked repo is returning data for the original repo instead.  Would you be able to confirm this?  You should see a GET for `/repos/personal-fork/repo-name` in the Network tab, but the response contents will describe the original repo.  (Or perhaps it redirects the request -- it would be interesting to know which, and whether any reason is given.)

If this is indeed the case I could probably work around it, but it's likely to break a bunch of other things besides posting comments.  I also have no idea how to reproduce this:  AFAIK it's very much not normal behavior, whether on github.com or GHE.  If you confirm I'll engage GitHub support to help me figure out why this is happening.  Thanks.

**Comment by N-Kass** [comment]
The GET to /repos is to my fork, and it's return value seems valid (forked repo data, not parent repo data).
However, the collaborators are from the original repo.
I think this actually makes sense, since my collaborators are members of the original repo, not of my repo (yet I want them to review my PR in my repo).

The GET to /repos/ also contained the original repo collaborators url, under "parent" & "source". 
The GET to /collaborators/ was straight to the original repo, not my forked one.

**Comment by pkaminski** [comment]
Something doesn't make sense here. Could you please send me a copy of the
response to the /repos fork fetch? Either unredacted or with consistent
replacements for sensitive data.  support@reviewable.io. Thanks!

**Comment by N-Kass** [comment]
Sent unredacted to support mail

**Comment by pkaminski** [comment]
Thanks so much -- that explained everything!  I didn't expect the structure for a personal repo to have an `organization` section.  This will be fixed in the next enterprise release.

**Comment by artiomle** [comment]
Hi Piotr,
Thanks for update
could you please share with me when next version will be released so I could download the images and update our system?

________________________________

**Comment by pkaminski** [comment]
Yep, I'll update (and close) this issue when the release is ready.

**Comment by pkaminski** [comment]
This should be resolved in v2351.4070, released today.  Please let me know if you're still having issues.
